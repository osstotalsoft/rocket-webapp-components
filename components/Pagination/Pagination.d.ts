import { TablePaginationProps } from '@material-ui/core/TablePagination'

export interface Props {

    /**
     * The total number of pages. This property is required
     */
    totalCount: number;

    /**
     * The number of items to be displayed in one page. This property is required
     */
    pageSize: number;

    /**
     * The current page. This property is required
     */
    page: number;

    /**
     * This function is called when the number of rows per page is changed
     * @param {object} event The event source of the callback.
     */
    onRowsPerPageChange?: (event: React.SyntheticEvent) => void;

    /**
     * This function is called when the page is changed
     * @param {object} event The event source of the callback.
     * @param {number} newPage The page you want to navigate
     * @param {number} direction 0 => navigate to a previous page and 1 => navigate to a next page
     */
    onPageChange?: (event: React.SyntheticEvent, newPage: number, direction: 0 | 1) => void;

    /**
    * @default () => { } 
    * The function called on refresh. Thi property is required
     
     */
    onRefresh: () => void;

    /**
     * If the pages are loading, thi property is set to true
     */
    loading?: boolean;

    /**
     * @default [10, 25, 50, 100] 
    * Array of values to choose how many rows should be displayed in a page
     */
    rowsPerPageOptions?: number[];

    /**
    * @default "of" 
    * The text to be displayed between the number of rows per page and total count
     */
    rowsOfText?: string;

    /**
      * @default "Rows per page" 
    * The text to be displayed for the number of rows per page
     */
    rowsPerPageText?: string;

    /**
     * Demos:
     * 
     * - https://bit.dev/totalsoft_oss/react-mui/pagination
     * 
     *  Take a look over the code for further information:https://github.com/osstotalsoft/rocket-webapp-components/blob/master/components/Pagination/Pagination.js
     * 
     * Material-UI components:
     * 
     * - TablePagination: https://material-ui.com/api/table-pagination/
     * 
     */

}

export type PaginationProps = Props | TablePaginationProps

export default function Pagination(props: PaginationProps): JSX.Element;