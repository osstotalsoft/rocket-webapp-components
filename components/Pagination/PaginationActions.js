import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { IconButton } from '@material-ui/core';
import KeyboardArrowLeft from '@material-ui/icons/KeyboardArrowLeft';
import KeyboardArrowRight from '@material-ui/icons/KeyboardArrowRight';

function TablePaginationActions({ count, page, rowsPerPage, onPageChange }) {

    const handleBackButtonClick = useCallback(event => {
        onPageChange(event, page - 1, 0);
    }, [onPageChange, page])

    const handleNextButtonClick = useCallback(event => {
        onPageChange(event, page + 1, 1);
    }, [onPageChange, page])

    return (
        <div style={{ display: "flex" }}>
            <IconButton
                onClick={handleBackButtonClick}
                disabled={page === 0}
                aria-label="Previous Page"
            >
                <KeyboardArrowLeft />
            </IconButton>
            <IconButton
                onClick={handleNextButtonClick}
                disabled={page >= Math.ceil(count / rowsPerPage) - 1}
                aria-label="Next Page"
            >
                <KeyboardArrowRight />
            </IconButton>
        </div>
    );
}

TablePaginationActions.propTypes = {
    count: PropTypes.number.isRequired,
    onPageChange: PropTypes.func.isRequired,
    page: PropTypes.number.isRequired,
    rowsPerPage: PropTypes.number.isRequired
};

export default TablePaginationActions;