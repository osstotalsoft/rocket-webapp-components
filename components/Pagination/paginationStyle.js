import styles from '../../assets/jss/styles.js';

const paginationStyle = theme => {
    const { defaultFont } = styles(theme);

    return {
        floatRight: {
            float: "right"
        },
        refreshButton: {
            lineHeight: "50px",
            float: "right",
            marginRight: "10px"
        },
        nextButton: {
            lineHeight: "50px",
            float: "right",
            marginRight: "10px",
            marginLeft: "10px"
        },
        selectRoot: {
            margin: "0!important"
        },
        actions: {
            margin: "0px"
        },
        root: {
            margin: "10px"
        },
        select: {
            ...defaultFont
        }
    };
}

export default paginationStyle;