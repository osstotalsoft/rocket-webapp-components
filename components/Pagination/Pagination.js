import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import { TablePagination, makeStyles } from '@material-ui/core';
import paginationStyle from './paginationStyle';
import TablePaginationActions from './PaginationActions';
import RefreshIcon from "@material-ui/icons/Refresh";
import { IconButton } from '@bit/totalsoft_oss.react-mui.kit.core';

const useStyles = makeStyles(paginationStyle);

const displayedRows = rowsOfText => ({ from, to, count }) => {
  return `${from}-${to} ${rowsOfText} ${count}`;
}

export const Pagination = ({ loading, totalCount, pageSize, page, rowsPerPageOptions, onRefresh,
  onPageChange, onRowsPerPageChange, rowsPerPageText, rowsOfText }) => {
  const classes = useStyles();

  const internalChangePage = useCallback((_event, newPage, direction) => {
    onPageChange(newPage, direction);
  }, [onPageChange])

  const internalChangePageSize = useCallback(event => {
    onRowsPerPageChange(event.target.value);
  }, [onRowsPerPageChange])

  const actualRowsPerPage = rowsPerPageOptions ? rowsPerPageOptions : [10, 25, 50, 100];

  const handleRefresh = useCallback(() => onRefresh(), [onRefresh])

  return !loading && (
    <div>
      <div className={classes.floatRight}>
        <TablePagination
          ActionsComponent={TablePaginationActions}
          component="div"
          count={totalCount}
          rowsPerPage={pageSize}
          page={page}
          onPageChange={internalChangePage}
          onRowsPerPageChange={internalChangePageSize}
          labelRowsPerPage={rowsPerPageText}
          labelDisplayedRows={displayedRows(rowsOfText)}
          SelectProps={{
            inputProps: { 'aria-label': 'Rows per page' },
            native: true
          }}
          rowsPerPageOptions={actualRowsPerPage}
        />
      </div>
      <div className={classes.refreshButton}>
        {onRefresh && <IconButton onClick={handleRefresh} color="defaultNoBackground" size='small'>
          <RefreshIcon />
        </IconButton >
        }
      </div>
    </div>
  );
}

Pagination.defaultProps = {
  onRefresh: () => { },
  rowsOfText: "of",
  rowsPerPageText: "Rows per page",
}

Pagination.propTypes = {
  /**
  * The total number of pages. This property is required
  */
  totalCount: PropTypes.number.isRequired,

  /**
  * The number of items to be displayed in one page. This property is required
  */
  pageSize: PropTypes.number.isRequired,
  /**
  * The current page. This property is required
  */
  page: PropTypes.number.isRequired,
  /**
  * This function is called when the number of rows per page is changed
  * @param {object} event The event source of the callback.
  */
  onRowsPerPageChange: PropTypes.func,
  /**
  * This function is called when the page is changed
  * @param {object} event The event source of the callback.
  * @param {number} newPage The page you want to navigate
  * @param {number} direction 0 => navigate to a previous page and 1 => navigate to a next page
  */
  onPageChange: PropTypes.func,
  /**
  * The function called on refresh. Thi property is required 
  */
  onRefresh: PropTypes.func.isRequired,
  /**
  * If the pages are loading, thi property is set to true
  */
  loading: PropTypes.bool,
  /**
  * Array of values to choose how many rows should be displayed in a page
  */
  rowsPerPageOptions: PropTypes.array,
  /**
  * The text to be displayed between the number of rows per page and total count
  */
  rowsOfText: PropTypes.string,
  /**
  * The text to be displayed for the number of rows per page
  */
  rowsPerPageText: PropTypes.string
};

export default Pagination;