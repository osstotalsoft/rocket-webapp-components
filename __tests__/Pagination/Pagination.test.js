import React from "react";
import renderer from "react-test-renderer";
import Pagination from "../../components/Pagination/Pagination";

describe("Pagination", () => {
  it("Snapshot", () => {
    const component = renderer.create(
      <Pagination
        totalCount={100}
        pageSize={100}
        page={0}
        onRowsPerPageChange={() => { }}
        onPageChange={() => { }}
        onRefresh={() => { }}
        loading={false}
      />
    );

    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
