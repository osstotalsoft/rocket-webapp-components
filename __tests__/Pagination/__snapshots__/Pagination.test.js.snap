// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Snapshot 1`] = `
<div>
  <div
    className="makeStyles-floatRight-1"
  >
    <div
      className="MuiTablePagination-root"
    >
      <div
        className="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
      >
        <div
          className="MuiTablePagination-spacer"
        />
        <p
          className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
        >
          Rows per page
        </p>
        <div
          className="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
          onClick={[Function]}
        >
          <select
            aria-label="Rows per page"
            className="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiInputBase-input"
            multiple={false}
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            value={100}
          >
            <option
              className="MuiTablePagination-menuItem"
              value={10}
            >
              10
            </option>
            <option
              className="MuiTablePagination-menuItem"
              value={25}
            >
              25
            </option>
            <option
              className="MuiTablePagination-menuItem"
              value={50}
            >
              50
            </option>
            <option
              className="MuiTablePagination-menuItem"
              value={100}
            >
              100
            </option>
          </select>
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
        </div>
        <p
          className="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
        >
          1-100 of 100
        </p>
        <div
          style={
            Object {
              "display": "flex",
            }
          }
        >
          <button
            aria-label="Previous Page"
            className="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 16.59L10.83 12l4.58-4.59L14 6l-6 6 6 6 1.41-1.41z"
                />
              </svg>
            </span>
          </button>
          <button
            aria-label="Next Page"
            className="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M8.59 16.59L13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="makeStyles-refreshButton-2"
  >
    <button
      className="MuiButtonBase-root MuiIconButton-root makeStyles-button-7 makeStyles-defaultNoBackground-15"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M17.65 6.35C16.2 4.9 14.21 4 12 4c-4.42 0-7.99 3.58-7.99 8s3.57 8 7.99 8c3.73 0 6.84-2.55 7.73-6h-2.08c-.82 2.33-3.04 4-5.65 4-3.31 0-6-2.69-6-6s2.69-6 6-6c1.66 0 3.14.69 4.22 1.78L13 11h7V4l-2.35 2.35z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
